vm/src/builtins/type.rs

```diff
@@ -913,13 +913,41 @@ impl PyType {
         Ok(())
     }
 
+    #[pygetset(magic)]
+    fn doc(&self, vm: &VirtualMachine) -> Option<PyObjectRef> {
+        if !self.slots.flags.has_feature(PyTypeFlags::HEAPTYPE) {
+            return self
+                .slots
+                .doc
+                .and_then(|doc| get_doc_from_internal_doc(&self.name(), doc))
+                .map(|signature| signature.to_pyobject(vm));
+        }
+
+        self.attributes
+            .read()
+            .get(identifier!(vm, __doc__))
+            .cloned()
+    }
+
     #[pygetset(magic)]
     fn text_signature(&self) -> Option<String> {
         self.slots
             .doc
             .and_then(|doc| get_text_signature_from_internal_doc(&self.name(), doc))
             .map(|signature| signature.to_string())
     }
+
+    #[pygetset(magic, setter)]
+    fn set_doc(&mut self, value: PyObjectRef, vm: &VirtualMachine) -> PyResult<()> {
```

Comment:
```suggestion
    fn set_doc(&self, value: PyObjectRef, vm: &VirtualMachine) -> PyResult<()> {
```
`mut` here is not allowed. `self.slots.doc` needs a lock.